vm/devices/storage/nvme/src/queue.rs

```diff
@@ -3,266 +3,342 @@
+        }
+        self.mem = mem;
+        self.offset = offset;
+        self.event_idx_offset = event_idx_offset;
+        Ok(())
     }
 
-    pub fn read(&self) -> u32 {
-        self.value.load(Ordering::SeqCst)
+    pub fn try_write(&self, db_id: u16, value: u32) -> Result<(), InvalidDoorbell> {
+        if (db_id as usize) >= self.wakers.len() {
+            return Err(InvalidDoorbell);
+        }
+        self.write(db_id, value);
+        Ok(())
     }
 
-    pub async fn wait_read(&self, value: u32) -> u32 {
-        let v = self.read();
-        if value != v {
-            return v;
+    fn write(&self, db_id: u16, value: u32) {
+        assert!((db_id as usize) < self.wakers.len());
+        let addr = self
+            .offset
+            .wrapping_add((db_id as u64) << DOORBELL_STRIDE_BITS);
+        if let Err(err) = self.mem.write_plain(addr, &value) {
+            tracelimit::error_ratelimited!(
+                error = &err as &dyn std::error::Error,
+                "failed to write doorbell memory"
+            );
         }
-        loop {
-            let listener = self.event.listen();
-            let v = self.read();
-            if value != v {
-                break v;
-            }
-            listener.await;
+        if let Some(waker) = &self.wakers[db_id as usize] {
+            waker.wake_by_ref();
+        }
+    }
+
+    fn read(&self, db_id: u16) -> Option<u32> {
+        assert!((db_id as usize) < self.wakers.len());
+        self.mem
+            .read_plain(
+                self.offset
+                    .wrapping_add((db_id as u64) << DOORBELL_STRIDE_BITS),
+            )
+            .inspect_err(|err| {
+                tracelimit::error_ratelimited!(
+                    error = err as &dyn std::error::Error,
+                    "failed to read doorbell memory"
+                );
+            })
+            .ok()
+    }
+
+    fn has_event_idx(&self) -> bool {
+        self.event_idx_offset.is_some()
+    }
+
+    fn write_event_idx(&self, db_id: u16, val: u32) {
+        assert!((db_id as usize) < self.wakers.len());
+        if let Err(err) = self.mem.write_plain(
+            self.event_idx_offset
+                .unwrap()
+                .wrapping_add((db_id as u64) << DOORBELL_STRIDE_BITS),
+            &val,
+        ) {
+            tracelimit::error_ratelimited!(
+                error = &err as &dyn std::error::Error,
+                "failed to read event_idx memory"
+            )
         }
     }
+
+    fn read_event_idx(&self, db_id: u16) -> Option<u32> {
+        assert!((db_id as usize) < self.wakers.len());
+        self.mem
+            .read_plain(
+                self.event_idx_offset?
+                    .wrapping_add((db_id as u64) << DOORBELL_STRIDE_BITS),
+            )
+            .inspect_err(|err| {
+                tracelimit::error_ratelimited!(
+                    error = err as &dyn std::error::Error,
+                    "failed to read doorbell memory"
```

Comment:
nit: add some identifier so that it's possible to debug which nvme device is seeing this error (here and in other places).

Reply:
I think this should probably be handled via a parent span or something rather than per-trace-point fields. Which we should do, but I'm not going to do in this PR.